docs: fully revised lint configuration guide (#517)
Redocly/openapi-cli: docs/guides/lint.md

---

# Introduction

The `lint` configuration section is part of the [Redocly configuration file](../configuration/configuration-file.mdx).
The `lint` and `bundle` commands use this section to control various options.

The `lint` configuration section consists of several sub-sections. The following code block shows an example `lint` section. Sub-sections are described further in the text.

```yaml
lint:
  plugins:
    - './local-plugin.js'
  extends:
    - recommended
  resolve:
    http:
      headers:
        - matches: https://api.example.com/v2/**
          name: X-API-KEY
          value: <direct value>
          envVariable: <name of env variable to be used as value>
        - matches: https://example.com/*/test.yaml
          name: Authorization
          value: <direct value>
          envVariable: <name of env variable to be used as value>
  rules:
    no-sibling-refs:
      severity: error
    boolean-parameter-prefixes:
      severity: error
      prefixes: ['should', 'is', 'has']
  ...
```

## Sub-sections

### Plugins

Use this section to import local plugins. If you don't have any custom plugins, omit this section.

* **type**: `array of strings`

> :::warning
> You don't need to import built-in plugins and rules.
> 
> Community plugins are not supported.
> :::
> 

Reviewer: ```suggestion
:::warning

You don't need to import built-in plugins and rules.

Community plugins are not supported.

:::

```

#### Examples

```yaml Import a single plugin
lint:
  plugins:
    - './local-plugin.js'
```

```yaml Import multiple plugins
lint:
  plugins:
    - ['./local-plugin.js', './another-local-plugin.js']
```

### Extends

Use this section to choose the base configuration for further extension or adding your own plugins. You may override specific settings in the subsequent sections.

* **type**: `array of strings`
* **default**: `recommended`
* **possible values**: `minimal`, `recommended`, `all`

> :::info

Reviewer: ```suggestion
```

Possible values specified above are related to built-in rules only. Custom plugins can contain additional configurations, so it can be as follows:

```yaml
lint:
  extends:
    - recommended
    - my-custom-super-ruleset
```

Find more information in the [Configs in plugins](../resources/custom-rules/#configs-in-plugins) section.

> :::

Reviewer: ```suggestion
```

    - minimal
```

```yaml Multiple values
lint:
  extends:
    - minimal
    - recommended
```

### Resolve

Use this section to specify external links in your definition that are not publicly accessible (except for Redocly API registry links).

* **type**: `object`

Redocly automatically resolves any API registry link or public URL. However, you may want to resolve links that are not API registry links or publicly accessible.

Currently, OpenAPI CLI only supports `http` headers. Only one `http` header per URL is supported in the `resolve` section:

```yaml
lint:
  resolve:
    http:
      headers:
        - # header configuration
```

#### Header configuration

| Property      | Description | Examples |
| ------------- | ----------- | -------- |
| `matches`     | Glob match against the URL. | `https://api.example.com/**` or `https://api.example.com/*.json` |
| `name`        | HTTP header name.           | `Authorization` or `X-API-KEY` |
| `value`       | The value of the header. Mutually exclusive with `envVariable`. | `123-abc` |
| `envVariable` | The name of the environment variable that contains the value of the header. Mutually exclusive with `value`. | `SECRET_KEY` |

> :::success
> It is recommended to use environment variables where possible.
> :::

Reviewer: ```suggestion
:::success

It is recommended to use environment variables where possible.

:::
```

#### Example

```yaml
lint:
  resolve:
    http:
      headers:
        - matches: https://api.example.com/v2/**
          name: X-API-KEY
          envVariable: SECRET_KEY
        - matches: https://example.com/*/test.yaml
          name: Authorization
          envVariable: SECRET_AUTH
```

The first match takes precedence when a URL matches multiple patterns. Therefore, only the header from the first match will be used in the request.

### Preprocessors

> As preprocessors are rarely indicated, omit this section.

Reviewer: ```suggestion
As preprocessors are rarely indicated, you can omit this section in most cases.
```

* **type**: `array of objects`
* **possible values**: `error`, `warn`, `off`

Preprocessors run first during `lint` and `bundle`.

:::info

> Linting happens only when the `--lint` flag is supplied

Reviewer: ```suggestion
Linting happens only when the `--lint` flag is used with the `bundle` command.
```

> :::info
> For the `bundle` command, linting happens only when the `--lint` flag is used.
> :::

Reviewer: ```suggestion
:::info

For the `bundle` command, linting happens only when the `--lint` flag is used.

:::
```

* **type**: `array of objects`
* **possible values**: `error`, `warn`, `off`

For `lint` command: rules run *after* preprocessors.
For `bundle` command: rules run *between* preprocessors and decorators.

#### Examples

```yaml Short syntax
lint:
  rules:
    no-sibling-refs: error
```

```yaml Verbose syntax
lint:
  rules:
    no-sibling-refs:
      severity: error
```

```yaml Rules with additional configuration
# Use verbose configuration syntax to define additional configuration
# The boolean-parameter-prefixes example overrides the default "prefixes".
lint:
  rules:
    boolean-parameter-prefixes:
      severity: error
      prefixes: ['should', 'is', 'has']
```    

### Decorators

Use this section to enable or disable decorators. They modify the definition in the bundling process after validation is complete. If you don't use decorators, omit this section.

* **type**: `array of objects`
* **possible values**: `error`, `warn`, `off`

For `bundle` command: decorators run *after* linting.

> :::info
> Linting happens only when the `--lint` flag is supplied
> :::

Reviewer: ```suggestion
:::info

For the `bundle` command, linting happens only when the `--lint` flag is used.

:::
```

  rules:
    no-sibling-refs:
      severity: error
    boolean-parameter-prefixes:
      severity: error
      prefixes: ['should', 'is', 'has']
  ...
```

## Sub-sections

### Plugins

Use this section to import local plugins. If you don't have any custom plugins, omit this section.

* **type**: `array of strings`

:::warning

You don't need to import built-in plugins and rules.

Community plugins are not supported.

:::

#### Examples

```yaml Import a single plugin
lint:
  plugins:
    - './local-plugin.js'
```

```yaml Import multiple plugins
lint:
  plugins:
    - ['./local-plugin.js', './another-local-plugin.js']
```

### Extends

Use this section to choose the base configuration for further extension or adding your own plugins. You may override specific settings in the subsequent sections.

* **type**: `array of strings`
* **default**: `recommended`
* **possible values**: `minimal`, `recommended`, `all`

Possible values specified above are related to built-in rules only. Custom plugins can contain additional configurations, so it can be as follows:

```yaml
lint:
  extends:
    - recommended
    - my-custom-super-ruleset
```

> Find more information in the [Configs in plugins](../resources/custom-rules/#configs-in-plugins) section.

Reviewer 2: Not to self -- we may need to adjust this to a friendly format for the link checker.

The `lint` and `bundle` commands use this section to control various options.

The `lint` configuration section consists of several sub-sections (each section is described after this block):

```yaml
lint:
  plugins:
    - './local-plugin.js'
  extends:
    - recommended
  resolve:
    http:
      headers:
        - matches: https://api.example.com/v2/**
          name: X-API-KEY
          value: <direct value>
          envVariable: <name of env variable to be used as value>
        - matches: https://example.com/*/test.yaml
          name: Authorization
          value: <direct value>
          envVariable: <name of env variable to be used as value>
  rules:
    no-sibling-refs:
      severity: error
    boolean-parameter-prefixes:
      severity: error
      prefixes: ['should', 'is', 'has']
  ...
```

## Sub-sections

### Plugins

Use this section to import local plugins.

* **type**: `array of strings`

:::warning
You don't need to import built-in plugins and rules.

Community plugins are not supported.
:::

:::success
Omit this section if you don't have custom plugins.
:::

#### Examples

> ```yaml single value

Reviewer: ```suggestion
```yaml Import a single plugin
```

lint:
  plugins:
    - './local-plugin.js'
```

> ```yaml multiple values

Reviewer: ```suggestion
```yaml Import multiple plugins
```

lint:
  plugins:
    - ['./local-plugin.js', './another-local-plugin.js']
```

### Extends

Use this section to choose the base configuration for further extension or adding your own plugins. You may override specific settings in the subsequent sections.

* **type**: `array of strings`
* **default**: `recommended`
* **possible values**: `minimal`, `recommended`, `all`

:::info
Possible values specified above are related to built-in rules only. Custom plugins can contain additional configurations, so it can be as follows:

```yaml
lint:
  extends:
    - recommended
    - my-custom-super-ruleset
```

Find more information in the [Configs in plugins](../resources/custom-rules/#configs-in-plugins) section.

:::

#### Examples

> ```yaml single value

Reviewer: ```suggestion
```yaml Single value
```

lint:
  extends:
    - minimal
```

> ```yaml multiple values

Reviewer: ```suggestion
```yaml Multiple values
```

lint:
  extends:
    - minimal
    - recommended
```

### Resolve

Use this section to specify external links in your definition that are not publicly accessible (except for Redocly API registry links).

* **type**: `object`

Redocly automatically resolves any API registry link or public URL. However, you may want to resolve links that are not API registry links or publicly accessible.

Currently, OpenAPI CLI only supports `http` headers. Only one `http` header per URL is supported in the `resolve` section:

```yaml
lint:
  resolve:
    http:
      headers:
        - # header configuration
```

#### Header configuration

| Property      | Description | Examples |
| ------------- | ----------- | -------- |

> | `matches`     | Glob match against the URL. | `https://api.example.com/**` or `https://api.example.com/*.json` |

Reviewer 3: query: is "glob" short for "global"? I would use **global** in that case, unless glob is really the word you were going for.

Author: Glob is the right word here. It is a pattern that matches files or directories or other pathnames based on wildcards (*, **).

| `name`        | HTTP header name.           | `Authorization` or `X-API-KEY` |
| `value`       | The value of the header. Mutually exclusive with `envVariable`. | `123-abc` |
| `envVariable` | The name of the environment variable that contains the value of the header. Mutually exclusive with `value`. | `SECRET_KEY` |

:::success

> It is recommeded to use environment variables where possible.

Reviewer: ```suggestion
It is recommended to use environment variables where possible.
```

:::

#### Example

> ```yaml multiple values

Reviewer: ```suggestion
```yaml
```

Reviewer: This custom text in code samples is visible only when you have two or more consecutive code blocks - that's when the portal displays them as tabs, with this text as the tab title. It won't be visible in this case, so we can remove it

lint:
  resolve:
    http:
      headers:
        - matches: https://api.example.com/v2/**
          name: X-API-KEY
          envVariable: SECRET_KEY
        - matches: https://example.com/*/test.yaml
          name: Authorization
          envVariable: SECRET_AUTH
```

> The first match will win in the event when a URL matches multiple patterns. Therefore, only the header from the first match will be used in the request.

Reviewer: ```suggestion
The first match takes precedence when a URL matches multiple patterns. Therefore, only the header from the first match will be used in the request.
```

### Preprocessors

> Use this section to change the [severity level](#severity-levels) of any rules in your extended configurations. Some rules may also receive [additional configurations](#additional-rule-options).

Reviewer: ```suggestion
Use this section to change the [severity level](#severity-levels) of any rules in your extended configurations. Some rules may also support [additional configurations](#additional-rule-options).
```

* **type**: `array of objects`
* **possible values**: `error`, `warn`, `off`

Preprocessors run first during `lint` and `bundle`.

:::info
Linting happens only when the `--lint` flag is supplied
:::

:::info

> Preprocessors are rarely indicated - avoid if possible.

Reviewer: question: what does "rarely indicated" mean in this case? I can't figure out if we're telling the users to avoid specifying any preprocessors, or to avoid using them altogether...?

This section can be omitted.

> :::

Reviewer 3: thought: Similar to the previous one, I would put this admonition before the actual section so that users can omit it if it does not apply to them.

Reviewer: :+1: 

Also, please consider what really needs to be an admonition and what can work as regular text.

### Rules

> Use this section to change the [severity level](#severity-levels) of any rules in your extended configurations. Some rules may also receive [additional configurations](#additional-rule-options).

Reviewer: issue: this text is identical as in the "Preprocessors" section - is that on purpose? Does that mean the Preprocessors and the Rules sections are used for the same thing?

Author: It seems like the text in the Preprocessors section is not relevant.

* **type**: `array of objects`
* **possible values**: `error`, `warn`, `off`

For `lint` command: rules run *after* preprocessors.
For `bundle` command: rules run *between* preprocessors and decorators.

#### Examples

> ```yaml short version

Reviewer: ```suggestion
```yaml Short syntax
```

lint:
  rules:
    no-sibling-refs: error
```

> ```yaml verbose version

Reviewer: ```suggestion
```yaml Verbose syntax
```

lint:
  rules:
    no-sibling-refs:
      severity: error
```

> ```yaml rules with additional configuration

Reviewer: ```suggestion
```yaml Rules with additional configuration
```

> # Use verbose configuration version to define additional configuration

Reviewer: ```suggestion
# Use verbose configuration syntax to define additional configuration
```

Reviewer: "style" could also work here if "syntax" is too technical. "Version" can be misleading so I would avoid it

# The boolean-parameter-prefixes example overrides the default "prefixes".
lint:
  rules:
    boolean-parameter-prefixes:
      severity: error
      prefixes: ['should', 'is', 'has']
```    

### Decorators

Use this section to enable or disable decorators. They modify the definition in the bundling process after validation is complete.

* **type**: `array of objects`
* **possible values**: `error`, `warn`, `off`

For `bundle` command: decorators run *after* linting.

:::info
Linting happens only when the `--lint` flag is supplied
:::

:::success

> Omit this section if you don't use decorators.

Reviewer: Same suggestion as in previous comments by @swapnilogale - let's move this further up in the text (and consider taking it out of the admonition box)

:::

### Severity levels

* **applied to**: [`preprocessors`](#preprocessors), [`rules`](#rules), [`decorators`](#decorators)
* **possible values**: `error`, `warn`, `off`

#### Examples

> ```yaml short version

Reviewer: ```suggestion
```yaml Short syntax
```

  - /docs/cli/configuration/lint/
---

> # Lint configuration

Reviewer 2: We start each article with a heading level 1. This helps our people understand what topic they are on, and helps search engines know about about the topic too.

The `lint` configuration section is part of the [Redocly configuration file](../configuration/configuration-file.mdx).
The `lint` and `bundle` commands use this section to control various options.

The `lint` configuration section consists of several sub-sections (each section is described after this block):

```yaml
lint:
  plugins:
    - './local-plugin.js'
  extends:
    - recommended
  resolve:
    http:
      headers:
        - matches: https://api.example.com/v2/**
          name: X-API-KEY
          value: <direct value>
          envVariable: <name of env variable to be used as value>
        - matches: https://example.com/*/test.yaml
          name: Authorization
          value: <direct value>
          envVariable: <name of env variable to be used as value>

>   preprocessors:

Reviewer 2: issue: I think these (preprocessors and decorators) sections aren't allowed to be empty. I know you let them know they can omit the section... but really, they should omit the section.

Author: So, is it better to remove both sections from the example completely?

Author: Removed both sections from the example

      severity: error
```

> :::info
> See the [rules documentation](../resources/built-in-rules.md) for more information.
> :::

Reviewer: ```suggestion
:::info

See the [rules documentation](../resources/built-in-rules.md) for more information.

:::
```

    no-unused-components: error
```

> ```yaml verbose version

Reviewer: ```suggestion
```yaml Verbose syntax
```

      severity: error
```

:::warning

> With the short version, you can't configure [additional options](#additional-rule-options) for any given rule (if it supports them).

Reviewer: ```suggestion
With the short configuration syntax, you can't configure [additional options](#additional-rule-options) for any given rule (if it supports them).
```

Reviewer: issue: This box should be moved up before we even show the short/verbose examples, so people know why they should use one over the other. Also, moving it up will resolve the problem of having two consecutive admonition boxes

    no-unused-components:
      severity: error
```

> To know which rules support options, please read the [built-in rules documentation](../resources/built-in-rules.md).

Reviewer 3: fix: remove the word "please" from docs

:::

### Different OpenAPI versions

> Redocly OpenAPI CLI supports OpenAPI of versions 2.0, 3.0, and 3.1. Most of the time you will use one of them. However, you may need to configure different rules based on the version. You can do that by using additional configuration sections:

Reviewer: ```suggestion
Redocly OpenAPI CLI supports OpenAPI versions 2.0, 3.0, and 3.1. Most of the time you will use one of them. However, you may need to configure different rules based on the version. You can do that by using additional configuration sections:
```

    boolean-parameter-prefixes: error
```

In this example, the OpenAPI specification version is identified:

* If it is version 2 (formerly known as Swagger), it will prioritize the `oas2Rules` section.
* If it is version 3 (OpenAPI 3.x), it will prioritize the `oas3_0Rules` section.

> :::info
> If the version is not defined, it will fall back to the `rules` section.
> :::

Reviewer: ```suggestion
:::info

If the version is not defined, it will fall back to the `rules` section.

:::
```
